### PR TITLE
MULE-10655 Cleanup MessageProcessorChains to ensure consistency of in…

### PR DESCRIPTION
**Pull request by dfeist**
…jection and lifecycle. REVIEW ONLY.

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java [hunk 1]
```diff
@@ -664,10 +667,10 @@ public void testMix() throws Exception {
   @Test
   public void testMixStaticFactoryt() throws Exception {
     MessageProcessorChain chain =
-        DefaultMessageProcessorChain.from(muleContext, new TestIntercepting(), new TestNonIntercepting(),
-                                          new TestNonIntercepting(), new TestIntercepting(), new TestNonIntercepting(),
-                                          new TestNonIntercepting());
-    Event restul = chain.process(getTestEventUsingFlow(""));
+        MessageProcessors.newChain(new TestIntercepting(), new TestNonIntercepting(),
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java]: static import

File: core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java [hunk 2]
```diff
@@ -689,19 +692,19 @@ public void testMix2() throws Exception {
   @Test
   public void testMix2StaticFactory() throws Exception {
     MessageProcessorChain chain =
-        DefaultMessageProcessorChain.from(muleContext, new TestNonIntercepting(), new TestIntercepting(),
-                                          new TestNonIntercepting(), new TestNonIntercepting(), new TestNonIntercepting(),
-                                          new TestIntercepting());
-    Event restul = chain.process(getTestEventUsingFlow(""));
+        MessageProcessors.newChain(new TestNonIntercepting(), new TestIntercepting(),
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java]: static import

File: core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java [hunk 3]
```diff
@@ -712,41 +715,37 @@ public void testResponseProcessor() throws MuleException, Exception {
...
-    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder(muleContext);
+  public void testResponseProcessorInNestedChain() throws Exception {
+    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder();
     final ResponseMessageProcessorAdapter responseMessageProcessorAdapter =
         new ResponseMessageProcessorAdapter(getAppendingMP("c"));
     responseMessageProcessorAdapter.setMuleContext(muleContext);
     builder.chain(
-                  getAppendingMP("1"), DefaultMessageProcessorChain.from(muleContext, getAppendingMP("a"),
-                                                                         responseMessageProcessorAdapter, getAppendingMP("b")),
+                  getAppendingMP("1"), MessageProcessors.newChain(getAppendingMP("a"),
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java]: static import

File: core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java [hunk 4]
```diff
@@ -783,53 +781,53 @@ public void testAll() throws MuleException, Exception {
...
     builder.chain(getAppendingMP("1"), new ExceptionThrowingMessageProcessor());
     process(builder.build(), getTestEventUsingFlow("0"));
   }
 
   @Test(expected = MessagingException.class)
-  public void testExceptionBefore() throws MuleException, Exception {
-    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder(muleContext);
+  public void testExceptionBefore() throws Exception {
+    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder();
     builder.chain(new ExceptionThrowingMessageProcessor(), getAppendingMP("1"));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java]: Maybe this chain() method can return the builder to use it fluently?

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 5]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
     }
   }
 
+  /**
+   * Sets an objects {@link FlowConstruct} it implements {@link FlowConstructAware}.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: *if* it implements

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 6]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
...
+   */
+  public static void setFlowConstructIfNeeded(Object object, FlowConstruct flowConstruct) {
+    object = unwrap(object);
+    if (object != null && object instanceof FlowConstructAware) {
+      ((FlowConstructAware) object).setFlowConstruct(flowConstruct);
+    }
+  }
+
+  /**
+   * Sets an objects {@link FlowConstruct} it implements {@link FlowConstructAware}.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: *if* it implements

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 7]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
...
+   * @param objects the object you're trying to set the flow construct
+   * @param flowConstruct the {@link FlowConstruct} in which the object is defined.
+   * @throws InitialisationException
+   */
+  public static void setFlowConstructIfNeeded(Collection<? extends Object> objects, FlowConstruct flowConstruct) {
+    objects.forEach(o -> setFlowConstructIfNeeded(o, flowConstruct));
+  }
+
+  /**
+   * Sets an objects {@link MuleContext} it implements {@link MuleContextAware}.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: *if* it implements

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 8]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
...
+   */
+  public static void setMuleContextIfNeeded(Object object, MuleContext muleContext) {
+    object = unwrap(object);
+    if (object != null && object instanceof MuleContextAware) {
+      ((MuleContextAware) object).setMuleContext(muleContext);
+    }
+  }
+
+  /**
+   * Sets an objects {@link MuleContext} it implements {@link MuleContextAware}.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: *if* it implements

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 9]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
...
+   *
+   * @param objects the object you're trying to set the mule context
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: the object*s* you're trying to set the mule context *to*

File: core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java [hunk 10]
```diff
@@ -270,6 +272,56 @@ public static void assertNotStopping(MuleContext muleContext, String errorMessag
...
+    object = unwrap(object);
+    if (object != null && object instanceof FlowConstructAware) {
+      ((FlowConstructAware) object).setFlowConstruct(flowConstruct);
+    }
+  }
+
+  /**
+   * Sets an objects {@link FlowConstruct} it implements {@link FlowConstructAware}.
+   *
+   * @param objects the object you're trying to set the flow construct
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/lifecycle/LifecycleUtils.java]: the object*s* you're trying to set the mule context *to*

File: core/src/main/java/org/mule/runtime/core/api/processor/MessageProcessors.java [hunk 11]
```diff
@@ -29,69 +22,58 @@ private MessageProcessors() {
...
+  }
 
-    @Override
-    public void setMuleContext(MuleContext context) {
-      if (delegate instanceof MuleContextAware) {
-        ((MuleContextAware) delegate).setMuleContext(context);
-      }
-    }
+  /**
+   * Creates a new explicit {@link MessageProcessorChain} from a {@link List} of {@link Processor}'s. Note that this performs
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/processor/MessageProcessors.java]: no List here, put *one or more* as in the previous method

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 12]
```diff
@@ -131,36 +133,11 @@ protected void initialiseProcessingStrategy() {
...
-
-      @Override
-      public Event process(Event event) throws MuleException {
-        muleContext.getNotificationManager()
-            .fireNotification(new PipelineMessageNotification(AbstractPipeline.this, event,
-                                                              PipelineMessageNotification.PROCESS_END));
-        return event;
-      }
-    });
+    builder.chain(new processEndProcessor());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: case

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 13]
```diff
@@ -309,23 +286,7 @@ public void addMessageProcessorPathElements(MessageProcessorPathElement pathElem
...
-          filteredMessageProcessorList.add(messageProcessor);
-          break;
-        }
-      } else {
-        filteredMessageProcessorList.add(messageProcessor);
-      }
-    }
-
-    NotificationUtils.addMessageProcessorPathElements(filteredMessageProcessorList, processorPathElement);
+    NotificationUtils.addMessageProcessorPathElements(pipeline, pathElement.addChild(processorsPrefix));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: static import

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 14]
```diff
@@ -386,4 +348,34 @@ protected void doDispose() {
     super.doDispose();
   }
 
+  private class processEndProcessor implements Processor, InternalMessageProcessor {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: first letter uppercase

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 15]
```diff
@@ -386,4 +348,34 @@ protected void doDispose() {
...
+
+    @Override
+    public Event process(Event event) throws MuleException {
+      muleContext.getNotificationManager()
+          .fireNotification(new PipelineMessageNotification(AbstractPipeline.this, event,
+                                                            PipelineMessageNotification.PROCESS_END));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: static import

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 16]
```diff
@@ -386,4 +348,34 @@ protected void doDispose() {
...
+
+
+  private class ProcessorStartCompleteProcessor extends AbstractRequestResponseMessageProcessor
+      implements InternalMessageProcessor {
+
+    @Override
+    protected Event processRequest(Event event) throws MuleException {
+      muleContext.getNotificationManager()
+          .fireNotification(new PipelineMessageNotification(AbstractPipeline.this, event,
+                                                            PipelineMessageNotification.PROCESS_START));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: static import

File: core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java [hunk 17]
```diff
@@ -386,4 +348,34 @@ protected void doDispose() {
...
+      return super.processRequest(event);
+    }
+
+    @Override
+    protected void processFinally(Event event, MessagingException exception) {
+      muleContext.getNotificationManager()
+          .fireNotification(new PipelineMessageNotification(AbstractPipeline.this, event,
+                                                            PipelineMessageNotification.PROCESS_COMPLETE, exception));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/AbstractPipeline.java]: static import

File: core/src/main/java/org/mule/runtime/core/construct/DynamicPipelineMessageProcessor.java [hunk 18]
```diff
@@ -71,23 +71,23 @@ private String resetAndUpdatePipeline(String id, List<Processor> preMessageProce
...
 
     // build new dynamic chains
-    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder(flow);
+    DefaultMessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder();
     builder.chain(preMessageProcessors);
 
     builder.chain(staticChain);
     builder.chain(postMessageProcessors);
-    MessageProcessorChain newChain = builder.build();
+    MessageProcessorChain from = builder.build();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/construct/DynamicPipelineMessageProcessor.java]: Is the variable name "from" correct?

File: core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java [hunk 19]
```diff
@@ -94,12 +96,8 @@ protected Event enrich(Event currentEvent,
...
 
   public void setEnrichmentMessageProcessor(Processor enrichmentProcessor) {
-    if (!(enrichmentProcessor instanceof MessageProcessorChain)) {
-      this.enrichmentProcessor = MessageProcessors.singletonChain(muleContext, enrichmentProcessor);
-      ((MuleContextAware) this.enrichmentProcessor).setMuleContext(muleContext);
-    } else {
-      this.enrichmentProcessor = enrichmentProcessor;
-    }
+    this.enrichmentProcessor = newChain(enrichmentProcessor);
+    ((MuleContextAware) this.enrichmentProcessor).setMuleContext(muleContext);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java]: Why not use the new method in LifecycleUtils here?

File: core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java [hunk 20]
```diff
@@ -159,11 +157,7 @@ public void setTarget(String target) {
 
   @Override
   public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
-    if (enrichmentProcessor instanceof InterceptingChainLifecycleWrapper) {
-      super.addMessageProcessorPathElements(pathElement);
-    } else {
-      ((MessageProcessorContainer) enrichmentProcessor).addMessageProcessorPathElements(pathElement);
-    }
+    NotificationUtils.addMessageProcessorPathElements(enrichmentProcessor, pathElement.addChild(this));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/enricher/MessageEnricher.java]: static import

File: core/src/main/java/org/mule/runtime/core/exception/AbstractExceptionListener.java [hunk 21]
```diff
@@ -340,4 +342,10 @@ void processOutboundRouterStatistics() {
       }
     }
   }
+
+  @Override
+  public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+    NotificationUtils.addMessageProcessorPathElements(messageProcessors, pathElement);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/exception/AbstractExceptionListener.java]: static import

File: core/src/main/java/org/mule/runtime/core/exception/TemplateOnErrorHandler.java [hunk 22]
```diff
@@ -176,13 +175,9 @@ protected Event route(Event event, MessagingException t) throws MessagingExcepti
...
     super.doInitialise(muleContext);
-    DefaultMessageProcessorChainBuilder defaultMessageProcessorChainBuilder =
-        new DefaultMessageProcessorChainBuilder(this.flowConstruct);
-    try {
-      configuredMessageProcessors = defaultMessageProcessorChainBuilder.chain(getMessageProcessors()).build();
-    } catch (MuleException e) {
-      throw new InitialisationException(e, this);
-    }
+    configuredMessageProcessors = newChain(getMessageProcessors());
+    configuredMessageProcessors.setFlowConstruct(flowConstruct);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/exception/TemplateOnErrorHandler.java]: LifecycleUtils

File: core/src/main/java/org/mule/runtime/core/processor/AsyncDelegateMessageProcessor.java [hunk 23]
```diff
@@ -86,14 +86,10 @@ public void initialise() throws InitialisationException {
       nameSource = ((StageNameSourceProvider) flowConstruct).getAsyncStageNameSource();
     }
 
-    MessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder(flowConstruct);
+    MessageProcessorChainBuilder builder = new DefaultMessageProcessorChainBuilder();
     processingStrategy.configureProcessors(Collections.singletonList(delegate), nameSource, builder,
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/AsyncDelegateMessageProcessor.java]: static import

File: core/src/main/java/org/mule/runtime/core/processor/AsyncDelegateMessageProcessor.java [hunk 24]
```diff
@@ -146,11 +142,7 @@ public ProcessingStrategy getProcessingStrategy() {
 
   @Override
   public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
-    if (delegate instanceof MessageProcessorContainer) {
-      ((MessageProcessorContainer) delegate).addMessageProcessorPathElements(pathElement);
-    } else {
-      NotificationUtils.addMessageProcessorPathElements(Collections.singletonList(delegate), pathElement);
-    }
+    NotificationUtils.addMessageProcessorPathElements(delegate, pathElement.addChild(this));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/AsyncDelegateMessageProcessor.java]: static import

File: core/src/main/java/org/mule/runtime/core/processor/chain/AbstractMessageProcessorChain.java [hunk 25]
```diff
@@ -165,10 +118,13 @@ public String toString() {
     return processors;
   }
 
+  protected List<Processor> getMessageProcessorsForLifecycle() {
+    return processors;
+  }
+
   @Override
   public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
     NotificationUtils.addMessageProcessorPathElements(getMessageProcessors(), pathElement);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/chain/AbstractMessageProcessorChain.java]: static import

File: core/src/main/java/org/mule/runtime/core/processor/chain/InternalMessageProcessorChainBuilder.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+  public static class InternalMessageProcessorChain extends DefaultMessageProcessorChain {
+
+    protected InternalMessageProcessorChain(String name, Processor head, List<Processor> processors,
+                                            List<Processor> processorsForLifecycle) {
+      super(name, head, processors, processorsForLifecycle);
+    }
+
+    @Override
+    public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+      NotificationUtils.addMessageProcessorPathElements(processors, pathElement);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/chain/InternalMessageProcessorChainBuilder.java]: static import

File: core/src/main/java/org/mule/runtime/core/processor/chain/SubflowMessageProcessorChainBuilder.java [hunk 27]
```diff
@@ -6,23 +6,62 @@
...
+    SubflowMessageProcessorChain(String name, Processor head, List<Processor> processors,
+                                 List<Processor> processorsForLifecycle) {
+      super(name, head, processors, processorsForLifecycle);
+      this.subFlowName = name;
+    }
+
+    @Override
+    public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+      MessageProcessorPathElement subprocessors = pathElement.addChild(name).addChild("subprocessors");
+      NotificationUtils.addMessageProcessorPathElements(processors, subprocessors);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/processor/chain/SubflowMessageProcessorChainBuilder.java]: static import

File: core/src/main/java/org/mule/runtime/core/routing/AbstractSelectiveRouter.java [hunk 28]
```diff
@@ -288,12 +292,14 @@ public void setRouterStatistics(RouterStatistics routerStatistics) {
...
-      messageProcessors.add(cmp.getMessageProcessor());
-    }
+    pathElement = pathElement.addChild(this);
+    List<Processor> messageProcessors =
+        conditionalMessageProcessors.stream().map(MessageProcessorFilterPair::getMessageProcessor).collect(toList());
     messageProcessors.add(defaultProcessor);
-    NotificationUtils.addMessageProcessorPathElements(messageProcessors, pathElement);
+    for (Processor route : messageProcessors) {
+      // Add additional child for each route, as the route represents the 'when' container in XML.
+      NotificationUtils.addMessageProcessorPathElements(route, pathElement.addChild(route));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/AbstractSelectiveRouter.java]: static import

File: core/src/main/java/org/mule/runtime/core/routing/AbstractSelectiveRouter.java [hunk 29]
```diff
@@ -192,7 +196,7 @@ public Event process(Event event) throws MuleException {
       return conditionalMessageProcessors;
     }
 
-    return ListUtils.union(conditionalMessageProcessors, Collections.singletonList(defaultProcessor));
+    return ListUtils.union(conditionalMessageProcessors, singletonList(defaultProcessor));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/AbstractSelectiveRouter.java]: static import

File: core/src/main/java/org/mule/runtime/core/routing/ScatterGatherRouter.java [hunk 30]
```diff
@@ -326,4 +329,13 @@ public void setTimeout(long timeout) {
   public void setRoutes(List<Processor> routes) {
     this.routes = routes;
   }
+
+  @Override
+  public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+    pathElement = pathElement.addChild(this);
+    for (MessageProcessorChain route : routeChains) {
+      NotificationUtils.addMessageProcessorPathElements(route, pathElement);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/ScatterGatherRouter.java]: static import

File: core/src/main/java/org/mule/runtime/core/routing/ScatterGatherRouter.java [hunk 31]
```diff
@@ -289,14 +292,14 @@ public void removeRoute(Processor processor) throws MuleException {
     routes.remove(processor);
   }
 
-  private void buildRouteChains() throws MuleException {
+  private void buildRouteChains() {
     Preconditions.checkState(routes.size() > 1, "At least 2 routes are required for ScatterGather");
     routeChains = new ArrayList<>(routes.size());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/ScatterGatherRouter.java]: replace this with a stream()?

File: core/src/main/java/org/mule/runtime/core/routing/WireTap.java [hunk 32]
```diff
@@ -117,4 +119,9 @@ public String toString() {
     return Collections.singletonList(tap);
   }
 
+  @Override
+  public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+    NotificationUtils.addMessageProcessorPathElements(getOwnedMessageProcessors(), pathElement.addChild(this));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/WireTap.java]: static import

File: core/src/main/java/org/mule/runtime/core/routing/outbound/AbstractOutboundRouter.java [hunk 33]
```diff
@@ -311,4 +313,10 @@ public void setFlowConstruct(FlowConstruct flowConstruct) {
     super.setFlowConstruct(flowConstruct);
     notificationTemplate.setFlowConstruct(flowConstruct);
   }
+
+  @Override
+  public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+    NotificationUtils.addMessageProcessorPathElements(getOwnedMessageProcessors(), pathElement.addChild(this));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/routing/outbound/AbstractOutboundRouter.java]: static import

File: core/src/main/java/org/mule/runtime/core/util/NotificationUtils.java [hunk 34]
```diff
@@ -76,29 +78,46 @@ public String resolvePath(Processor processor) {
...
+   * Adds the processor path for a specific {@link Processor}'s to the provided parent {@link MessageProcessorPathElement} as a
+   * child. {@link InternalMessageProcessor} implementations are ignored. Where a {@link Processor} implements
+   * {@link MessageProcessorContainer} it will not automatically be added as a child here, but rather this logic and the decision
+   * regarding what child elements should be added is deletagted to the {@link MessageProcessorContainer} via it's
+   * {@link MessageProcessorContainer#addMessageProcessorPathElements(MessageProcessorPathElement)} method.
+   *
+   * @param processor the {@link Processor} whose path is to be added to the parent element.
+   * @param parentElement the parent {@link MessageProcessorPathElement}
+   */
+  public static void addMessageProcessorPathElements(Processor processor,
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/util/NotificationUtils.java]: Is there an issue created to handle this logic at the parser level?

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBean.java [hunk 35]
```diff
@@ -134,11 +140,16 @@ public void dispose() {
...
-      return createDynamicReferenceMessageProcessor(refName);
-    }
+    Processor processor =
+        referencedMessageProcessor != null ? referencedMessageProcessor : createDynamicReferenceMessageProcessor(refName);
+    // Wrap in chain to ensure the flow-ref element always has a path element and lifecycle will be propgated to child sub-flows
+    return new AbstractMessageProcessorChain(singletonList(processor)) {
+
+      @Override
+      public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+        NotificationUtils.addMessageProcessorPathElements(processor, pathElement.addChild(processor));
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBean.java]: static import

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBean.java [hunk 36]
```diff
@@ -215,6 +226,11 @@ protected Processor lookupReferencedFlowInApplicationContext(String name) {
       return new FlowRefMessageProcessor() {
 
         @Override
+        public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+          NotificationUtils.addMessageProcessorPathElements(referencedFlow, pathElement);
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBean.java]: static import

File: modules/spring-config/src/test/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBeanTestCase.java [hunk 37]
```diff
@@ -53,16 +63,17 @@
...
   private ApplicationContext applicationContext = mock(ApplicationContext.class);
-  private MuleContext muleContext = mock(MuleContext.class);
   private ExpressionLanguage expressionLanguage = mock(ExpressionLanguage.class);
 
+  public FlowRefFactoryBeanTestCase() throws MuleException {}
+
   @Before
   public void setup() throws MuleException {
-    when(muleContext.getExpressionLanguage()).thenReturn(expressionLanguage);
+    muleContext.getRegistry().registerObject(MuleProperties.OBJECT_EXPRESSION_LANGUAGE, expressionLanguage);
```
Comment (elrodro83) [inline modules/spring-config/src/test/java/org/mule/runtime/config/spring/factories/FlowRefFactoryBeanTestCase.java]: static import

File: modules/ws/src/main/java/org/mule/runtime/module/ws/consumer/WSConsumer.java [hunk 38]
```diff
@@ -157,7 +158,10 @@ private Processor createMessageProcessor() throws MuleException {
 
     chainBuilder.chain(config.createOutboundMessageProcessor(flowConstruct));
 
-    return chainBuilder.build();
+    MessageProcessorChain chain = chainBuilder.build();
+    chain.setFlowConstruct(flowConstruct);
```
Comment (elrodro83) [inline modules/ws/src/main/java/org/mule/runtime/module/ws/consumer/WSConsumer.java]: LifecycleUtils

File: tests/integration/src/test/java/org/mule/test/core/context/notification/FlowStackTestCase.java [hunk 39]
```diff
@@ -199,7 +199,7 @@ public void subFlowStaticWithEnricher() throws Exception {
 
     assertStackElements(FlowStackAsserter.stackToAssert,
                         isFlowStackElement("subFlow", "/subFlowStaticWithEnricher/processors/0/0/subFlow/subprocessors/0"),
-                        isFlowStackElement("subFlowStaticWithEnricher", "/subFlowStaticWithEnricher/processors/0"));
+                        isFlowStackElement("subFlowStaticWithEnricher", "/subFlowStaticWithEnricher/processors/0/0"));
```
Comment (elrodro83) [inline tests/integration/src/test/java/org/mule/test/core/context/notification/FlowStackTestCase.java]: Is there a spec/issue to track this with debugger/studio?

File: transports/http/src/main/java/org/mule/compatibility/transport/http/components/HttpResponseBuilder.java [hunk 40]
```diff
@@ -342,5 +344,9 @@ public void setMessageProcessor(Processor messageProcessor) {
     ownedMessageProcessor.add(bodyTransformer);
   }
 
+  @Override
+  public void addMessageProcessorPathElements(MessageProcessorPathElement pathElement) {
+    NotificationUtils.addMessageProcessorPathElements(getOwnedMessageProcessors(), pathElement.addChild(this));
```
Comment (elrodro83) [inline transports/http/src/main/java/org/mule/compatibility/transport/http/components/HttpResponseBuilder.java]: static import

File: transports/module-support/src/test/java/org/mule/functional/AbstractScriptWithBindingsConfigBuilderTestCase.java [hunk 41]
```diff
@@ -60,9 +60,8 @@ public void testEndpointConfig() throws MuleException {
     assertNotNull(responseTransformers);
     assertFalse(responseTransformers.isEmpty());
     final Object responseTransformer = responseTransformers.get(0);
-    assertTrue(responseTransformer instanceof InterceptingChainLifecycleWrapper);
-    assertTrue(((InterceptingChainLifecycleWrapper) responseTransformer).getMessageProcessors()
-        .get(0) instanceof TestCompressionTransformer);
+    assertTrue(responseTransformer instanceof MessageProcessorChain);
```
Comment (elrodro83) [inline transports/module-support/src/test/java/org/mule/functional/AbstractScriptWithBindingsConfigBuilderTestCase.java]: hamcrest

File: core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java (hunk 4 above)
Comment (dfeist) [inline core-tests/src/test/java/org/mule/runtime/core/processor/chain/DefaultMessageProcessorChainTestCase.java]: It does, but i don't chain build() for readability.

File: core/src/main/java/org/mule/runtime/core/exception/TemplateOnErrorHandler.java (hunk 22 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/exception/TemplateOnErrorHandler.java]: Why, there is no cast here any more.

**Pull request merged**